### PR TITLE
List ethernets

### PR DESCRIPTION
**Pull request by ricab**
List ethernet devices in `multipass networks` when using the LXD backend. Part of the bridge creation story.

### CODE REVIEW

File: src/platform/platform_linux.cpp [hunk 1]
```diff
@@ -38,32 +38,103 @@
...
+
+    QFile type_file{net_dir.filePath(type_filename)};
+    if (type_file.open(QIODevice::ReadOnly | QIODevice::Text))
+    {
+        bool ok;
+        auto got = QTextStream{&type_file}.read(6).toInt(&ok); // 6 chars enough for up to 0xFFFF; 0 returned on failure
+        return ok ? got : default_ret;
+    }
+
+    mpl::log(mpl::Level::debug, category, fmt::format("Could not read {}", type_file.fileName()));
```
Comment (Saviq) [inline src/platform/platform_linux.cpp:72]:
```suggestion
    mpl::log(mpl::Level::warning, category, fmt::format("Could not read {}", type_file.fileName()));
```

Wonder if we should mention that `network-observe` needs to be connected to handle ethernet? Weird, though, that `network-control` doesn't give us that already…

File: src/platform/platform_linux.cpp [hunk 2]
```diff
@@ -38,32 +38,103 @@
...
+        QRegularExpression{QStringLiteral("^DEVTYPE=(.*)$"), QRegularExpression::MultilineOption};
+
+    QFile uevent_file{net_dir.filePath(uevent_filename)};
+    if (uevent_file.open(QIODevice::ReadOnly | QIODevice::Text))
+    {
+        auto contents = QTextStream{&uevent_file}.read(max_read);
+        return devtype_regex.match(contents).captured(1);
+    }
+
+    mpl::log(mpl::Level::debug, category, fmt::format("Could not read {}", uevent_file.fileName()));
```
Comment (Saviq) [inline src/platform/platform_linux.cpp:92]:
```suggestion
    mpl::log(mpl::Level::warning, category, fmt::format("Could not read {}", uevent_file.fileName()));
```

File: src/platform/platform_linux.cpp [hunk 3]
```diff
@@ -38,32 +38,103 @@
...
+    return {};
+}
+
+bool is_virtual_net(const QDir& net_dir)
+{
+    static const auto virtual_dir = QStringLiteral("virtual");
+
+    auto fixme = net_dir.canonicalPath();
```
Comment (Saviq) [inline src/platform/platform_linux.cpp:100]: ?

File: src/platform/platform_linux.cpp [hunk 4]
```diff
@@ -38,32 +38,103 @@
...
+    return net_dir.canonicalPath().contains(virtual_dir, Qt::CaseInsensitive);
+}
+
+bool is_ether_net(const QDir& net_dir)
```
Comment (Saviq) [inline src/platform/platform_linux.cpp:104]:
```suggestion
bool is_ethernet(const QDir& net_dir)
```

I'd say it's an Ethernet Network, not an Ether Network :]

**Review by Saviq — CHANGES_REQUESTED** [review]
Works!

Handful of comments inline.

File: snap/snapcraft.yaml [hunk 5]
```diff
@@ -44,6 +44,7 @@ apps:
       - multipass-support
       - network-bind
       - network-control
+      - network-observe # for network listing (to read `/sys/devices/**/net/*`)
```
[the remark this answers is not in the document]
Comment (ricab) [reply snap/snapcraft.yaml:47]: Requested: https://forum.snapcraft.io/t/auto-connect-network-interfaces-in-multipass/24060

File: src/platform/platform_linux.cpp (hunk 1 above)
Comment (ricab) [reply src/platform/platform_linux.cpp]: Good idea, will do. Yeah, `network-control` did not give us that.

Comment (ricab) [reply src/platform/platform_linux.cpp]: Done.

File: src/platform/platform_linux.cpp (hunk 2 above)
Comment (ricab) [reply src/platform/platform_linux.cpp]: Done.

File: src/platform/platform_linux.cpp (hunk 3 above)
Comment (ricab) [reply src/platform/platform_linux.cpp:102]: Heh thanks :sweat_smile:

File: src/platform/platform_linux.cpp (hunk 4 above)
Comment (ricab) [reply src/platform/platform_linux.cpp:106]: I suppose that's what ethernet means originally. I went for that for consistency with the other ones, and following the name of the constant (`ARPHRD_ETHER`). It is abbreviated like that in other places too (e.g. [EtherType](https://en.wikipedia.org/wiki/EtherType)). But ok, I don't mind changing it.

BTW, here's a history quote from [wikipedia](https://en.wikipedia.org/wiki/Ethernet#History):

> The idea was first documented in a memo that Metcalfe wrote on May 22, 1973, where he named it after the luminiferous aether once postulated to exist as an "omnipresent, completely-passive medium for the propagation of electromagnetic waves."

**Review by Saviq — APPROVED** [review]
Good!

Will let you merge when ready.

bors d+

**Comment by ricab** [comment]
Thanks!

bors r+

**Comment by ricab** [comment]
bors r=Saviq

**Comment by ricab** [comment]
bors r=Saviq

**Pull request merged**